### DISCUSSION TITLE
Relation of geospatial machine learning (GeoML) data science libraries

### DISCUSSION THREAD

**Discussion by weiji14** [weiji14/zen3geo#70] [General]
Inspired by @rbavery's blog post at https://developmentseed.org/blog/2022-09-27-ml-tooling, here's a table of GeoML Python libraries/frameworks focusing on the **Data Ops** side. This is a living document, so let me know if you see anything that needs updating!

|  | [zen3geo](https://github.com/weiji14/zen3geo/tree/v0.6.1) | [torchgeo](https://github.com/microsoft/torchgeo/tree/v0.4.1) | [eo-learn](https://github.com/sentinel-hub/eo-learn/tree/v1.4.1) | [raster-vision](https://github.com/azavea/raster-vision/tree/v0.20.2) |
|--|--|--|--|--|
| Year started | 2022 | 2021 | 2018 | 2018 |
| Organization | - | Microsoft | Sentinel-Hub/Sinergise [->](https://www.businesswire.com/news/home/20230329005285/en) Planet | Azavea [->](https://www.element84.com/blog/our-next-era-azavea-joins-element-84) Element84 |
| Version | v0.6.1 | v0.4.1 | v1.4.1 (production) | v0.20.2 |
| [Design](https://www.brandons.me/blog/libraries-not-frameworks) | [Lightweight library](https://github.com/weiji14/zen3geo/blob/v0.6.1/pyproject.toml#L32-L35) | [Heavyweight library](https://github.com/microsoft/torchgeo/blob/v0.4.1/setup.cfg#L27-L62) | [Library](https://github.com/sentinel-hub/eo-learn/tree/v1.4.1#installation)/[Framework](https://github.com/sentinel-hub/eo-learn/blob/v1.4.1/setup.py#L40-L49) | [Library](https://github.com/azavea/raster-vision/issues/1395)/[Framework](https://github.com/azavea/raster-vision/blob/v0.20.2/requirements.txt#L1-L6) |
| [Paradigm](https://en.wikipedia.org/wiki/Composition_over_inheritance) | Composition | Inheritance | Inheritance | Chained-inheritance |
| Data model | **xarray**/geopandas | [GeoDataset](https://torchgeo.readthedocs.io/en/v0.4.1/api/datasets.html#torchgeo.datasets.GeoDataset) (numpy/shapely) | [EOPatch](https://eo-learn.readthedocs.io/en/stable/examples/core/CoreOverview.html#EOPatch) (numpy/geopandas) | [Scene](https://docs.rastervision.io/en/0.20/usage/basics.html#scene) (numpy/geopandas) |
| STAC | [Yes](https://github.com/weiji14/zen3geo/discussions/48) | [No^](https://github.com/microsoft/torchgeo/issues/403) | No | [Yes](https://docs.rastervision.io/en/0.20/api_reference/_generated/rastervision.core.utils.stac.html) |
| Segmentation | Yes | Yes | Yes | Yes |
| Object Detection (bbox) | [Yes](https://zen3geo.readthedocs.io/en/v0.6.1/object-detection-boxes.html) | [DIY](https://github.com/microsoft/torchgeo/pull/458) | No? | [Yes](https://docs.rastervision.io/en/0.20/framework/examples.html#object-detection-cowc-potsdam-cars) |
| Multi-CRS without reprojection | [Yes](https://zen3geo.readthedocs.io/en/v0.6.1/chipping.html#pool-chips-into-mini-batches) | [No](https://github.com/microsoft/torchgeo/issues/278) | No? | No? |
| Multi-resolution | [Yes](https://zen3geo.readthedocs.io/en/v0.6.1/multi-resolution.html) via [datatree](https://github.com/xarray-contrib/datatree) + some DIY | [No](https://github.com/microsoft/torchgeo/issues/74) | [Yes](https://github.com/sentinel-hub/eo-learn/blob/v1.2.1/geometry/eolearn/geometry/superpixel.py) | No |
| Multi-dimensional (>2D+bands) | [Yes](https://zen3geo.readthedocs.io/en/v0.5.0/stacking.html), via xarray | No | No | No |
| Visualization | No | Yes | Yes | Yes |
| BYO custom function | [Easy](https://zen3geo.readthedocs.io/en/v0.6.1/vector-segmentation-masks.html#transform-and-visualize-raster-data) | [Hard](https://torchgeo.readthedocs.io/en/v0.4.1/tutorials/transforms.html) | [Hard](https://eo-learn.readthedocs.io/en/stable/examples/core/CoreOverview.html#EOTask) | [Hard](https://docs.rastervision.io/en/0.20/usage/tutorials/reading_raster_data.html#Transforming-rasters-using-RasterTransformers) |
| ML Library coupling | Pytorch or [None^](https://github.com/pytorch/data/issues/293) | Pytorch | None | Pytorch, [Tensorflow (DIY)](https://docs.rastervision.io/en/0.20/changelog.html#removed-tensorflow-backends) |

Key:
- `^`: coming
- `DIY` - Do it Yourself

Tying 'features' to 'applications':
- Multi-resolution - enables super-resolution, handle multimodal sensors (optical/SAR/gravity/etc)
- Multi-dimensional - enables time-series analysis, forecasting with physics-informed neural networks (climate models, hazard predictions) and beyond

Note that these extra 'features' come for free by following the [`xarray`](https://github.com/pydata/xarray) Data Model.

References:
 - [PyPI GIS/Image Processing libraries](https://pypi.org/search/?q=&o=-created&c=Topic+%3A%3A+Scientific%2FEngineering+%3A%3A+GIS&c=Topic+%3A%3A+Scientific%2FEngineering+%3A%3A+Image+Processing)
 - [PyPI GIS/Artificial Intelligence libraries](https://pypi.org/search/?q=&o=-created&c=Topic+%3A%3A+Scientific%2FEngineering+%3A%3A+Artificial+Intelligence&c=Topic+%3A%3A+Scientific%2FEngineering+%3A%3A+GIS)
